packages/caravan-wallets/src/jade.ts

```diff
@@ -0,0 +1,504 @@
+    );
+  }
+}
+
+export class JadeConfirmMultisigAddress extends JadeInteraction {
+  network: BitcoinNetwork;
+
+  bip32Path: string;
+
+  walletConfig: MultisigWalletConfig;
+
+  constructor({
+    network,
+    bip32Path,
+    walletConfig,
+  }: {
+    network: BitcoinNetwork;
+    bip32Path: string;
+    walletConfig: MultisigWalletConfig;
+  }) {
+    super();
+    this.network = network;
+    this.bip32Path = bip32Path;
+    this.walletConfig = walletConfig;
+  }
+
+  async run() {
+    return await this.withDevice(this.network, async (jade: IJade) => {
+      const descriptor = walletConfigToDescriptor(this.walletConfig);
+      let multisigName = await jade.getMultiSigName(this.network, descriptor);
+
+      if (!multisigName) {
+        multisigName = `jade${randomBytes(4).toString("hex")}`;
+        await jade.registerMultisig(this.network, multisigName, descriptor);
+      }
+      const wallet = await jade.getRegisteredMultisig(multisigName);
+      console.log("registered wallet", wallet);
+      const relativePath = this.bip32Path;
+
+      const paths = descriptor.signers.map((signer) => {
+        return extractPathSuffix(relativePath, signer.derivation);
+      });
+
+      const opts: ReceiveOptions = {
+        paths,
+        multisigName,
+      };
+
+      const multisigAddress = await jade.getReceiveAddress(this.network, opts);
+      return multisigAddress;
+    });
+  }
+}
+
+function parsePsbt(psbt: string): PsbtV2 {
+  const psbtVersion = getPsbtVersionNumber(psbt);
+  switch (psbtVersion) {
+    case 0:
+      return PsbtV2.FromV0(psbt, true);
+    case 2:
+      return new PsbtV2(psbt);
+    default:
+      throw new Error(`PSBT of unsupported version ${psbtVersion}`);
+  }
+}
+
+export class JadeSignMultisigTransaction extends JadeInteraction {
+  private walletConfig: MultisigWalletConfig;
+
+  private returnSignatureArray: boolean;
+
+  private unsignedPsbt: Uint8Array;
+
+  private base64string: string;
+
+  constructor({
+    walletConfig,
+    psbt,
+    returnSignatureArray = false,
+  }: {
+    walletConfig: MultisigWalletConfig;
+    psbt: string;
+    returnSignatureArray: boolean;
+  }) {
+    super();
+    this.walletConfig = walletConfig;
+    this.returnSignatureArray = returnSignatureArray;
+    this.unsignedPsbt = base64ToBytes(psbt);
+    this.base64string = psbt;
+  }
+
+  async run() {
+    return await this.withDevice(
+      this.walletConfig.network,
+      async (jade: IJade) => {
+        const signedPSBT = await jade.signPSBT(
+          this.walletConfig.network,
+          this.unsignedPsbt,
+        );
+
+        if (this.returnSignatureArray) {
+          const b64string = bytesToBase64(signedPSBT);
+          const parsedPsbt = parsePsbt(b64string);
+          const fingerprint = await jade.getMasterFingerPrint(
+            this.walletConfig.network,
+          );
+
+          let sigArray: string[] = [];
+
+          for (let i = 0; i < parsedPsbt.PSBT_GLOBAL_INPUT_COUNT; i++) {
```

Comment:
totally aesthetic comment, but this might be a nice utility to extract into a utility function at the top? keeps the interaction function more straightforward.